x-pack/platform/plugins/shared/streams/server/lib/streams/state_management/merge_unwired_stream_pipeline_actions.ts

```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { IngestPipeline } from '@elastic/elasticsearch/lib/api/types';
+import { isNotFoundError } from '@kbn/es-errors';
+import { castArray, groupBy, uniq } from 'lodash';
+import { IScopedClusterClient } from '@kbn/core/server';
+import { ASSET_VERSION } from '../../../../common/constants';
+import {
+  ActionsByType,
+  AppendProcessorToIngestPipelineAction,
+  DeleteProcessorFromIngestPipelineAction,
+  ElasticsearchAction,
+  UpsertWriteIndexOrRolloverAction,
+} from './execution_plan';
+
+const MANAGED_BY_STREAMS = 'streams';
+
+type UnwiredStreamPipelineAction =
+  | AppendProcessorToIngestPipelineAction
+  | DeleteProcessorFromIngestPipelineAction;
+
+// Should this be a static method on UnwiredStream instead?
+export async function mergeUnwiredStreamPipelineActions(
+  actionsByType: ActionsByType,
+  scopedClusterClient: IScopedClusterClient
+) {
+  const maybeActions = [
+    ...actionsByType.append_processor_to_ingest_pipeline,
+    ...actionsByType.delete_processor_from_ingest_pipeline,
+  ] as UnwiredStreamPipelineAction[];
+
+  if (maybeActions.length === 0) {
+    return;
+  }
+
+  const actionsByPipeline = groupBy(maybeActions, 'targetPipeline');
+
+  for (const [pipelineName, actions] of Object.entries(actionsByPipeline)) {
+    const pipeline = await getPipeline(pipelineName, scopedClusterClient);
+
+    if (pipeline) {
+      if (pipeline._meta?.managed_by === MANAGED_BY_STREAMS) {
+        updateExistingStreamsManagedPipeline({
+          pipelineName,
+          pipeline,
+          actions,
+          actionsByType,
+          scopedClusterClient,
+        });
+      } else {
+        updateExistingUserManagedPipeline({
+          pipelineName,
+          actions,
+          actionsByType,
+          scopedClusterClient,
+        });
+      }
+    } else {
+      createStreamsManagedPipeline({
+        actions,
+        actionsByType,
+        scopedClusterClient,
+      });
+    }
+  }
+
+  actionsByType.append_processor_to_ingest_pipeline = [];
+  actionsByType.delete_processor_from_ingest_pipeline = [];
+}
+
+async function createStreamsManagedPipeline({
+  actions,
+  actionsByType,
+  scopedClusterClient,
+}: {
+  actions: UnwiredStreamPipelineAction[];
+  actionsByType: ActionsByType;
+  scopedClusterClient: IScopedClusterClient;
+}) {
+  assertOnlyAppendActions(actions);
+  const targetTemplateNames = uniq(actions.map((action) => action.template));
+  if (targetTemplateNames.length !== 1) {
+    throw new Error('Append actions targeting the same new pipeline target different templates');
+  }
+  const indexTemplate = await getIndexTemplate(targetTemplateNames[0], scopedClusterClient);
+
+  const pipelineName = `${indexTemplate.name}@custom`;
```

Comment:
I'm not sure what to name this pipeline

Reply:
me neither... The `@custom` thing is there to disambiguate managed from non-managed stuff. It doesn't really apply if everything is user-owned. We could just do `${template-name}-pipeline`